File: python-flask/src/application/create_order_use_case.py
from src.application.dto import CreateOrderInput, CreateOrderOutput
from src.domain.order import Order, OrderItem
from src.domain.order_repository import OrderRepository


class CreateOrderUseCase:
    def __init__(self, repo: OrderRepository) -> None:
        self._repo = repo

    def execute(self, input_dto: CreateOrderInput) -> CreateOrderOutput:
        if not input_dto.customerId:
            raise ValueError("customerId is required")
        if not input_dto.items:
            raise ValueError("items must not be empty")

        items = []
        for i, item in enumerate(input_dto.items):
            if not item.productId:
                raise ValueError(f"item[{i}].productId is required")
            if item.quantity <= 0:
                raise ValueError(f"item[{i}].quantity must be greater than 0")
            if item.price < 0:
                raise ValueError(f"item[{i}].price must be >= 0")
            items.append(
                OrderItem(
                    product_id=item.productId,
                    quantity=item.quantity,
                    price=item.price,
                )
            )

        order = Order.create(input_dto.customerId, items)
        self._repo.save(order)

        return CreateOrderOutput(
            orderId=order.order_id,
            customerId=order.customer_id,
            totalAmount=order.total_amount,
            itemsCount=order.items_count,
            createdAt=order.created_at.strftime("%Y-%m-%dT%H:%M:%SZ"),
        )

File: nodejs-nestjs/src/application/create-order.use-case.ts
import { Injectable, Inject, BadRequestException } from '@nestjs/common';
import { CreateOrderInput, CreateOrderOutput } from './dto';
import { OrderItem, createOrder } from '../domain/order.entity';
import { OrderRepository, ORDER_REPOSITORY } from '../domain/order.repository.interface';

@Injectable()
export class CreateOrderUseCase {
  constructor(
    @Inject(ORDER_REPOSITORY) private readonly repo: OrderRepository,
  ) {}

  async execute(input: CreateOrderInput): Promise<CreateOrderOutput> {
    if (!input.customerId) {
      throw new BadRequestException('customerId is required');
    }
    if (!input.items || input.items.length === 0) {
      throw new BadRequestException('items must not be empty');
    }

    const items: OrderItem[] = input.items.map((item, i) => {
      if (!item.productId) {
        throw new BadRequestException(`item[${i}].productId is required`);
      }
      if (item.quantity <= 0) {
        throw new BadRequestException(`item[${i}].quantity must be greater than 0`);
      }
      if (item.price < 0) {
        throw new BadRequestException(`item[${i}].price must be >= 0`);
      }
      return new OrderItem(item.productId, item.quantity, item.price);
    });

    const order = createOrder(input.customerId, items);
    await this.repo.save(order);

    return {
      orderId: order.orderId,
      customerId: order.customerId,
      totalAmount: order.totalAmount,
      itemsCount: order.itemsCount,
      createdAt: order.createdAt,
    };
  }
}
